File: jikan-backend/src/lib/providers/animelok.ts

const USER_AGENT = 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/125.0.0.0 Safari/537.36';
const ANIMELOK_HEADERS = {
    'User-Agent': USER_AGENT,
    'Referer': 'https://animelok.xyz/',
    'Accept': 'application/json, text/plain, */*'
};

const anilistCache = new Map<string, string>();

export async function getAnilistId(malId: string): Promise<string | null> {
    if (anilistCache.has(malId)) return anilistCache.get(malId)!;
    try {
        const query = `query ($id: Int) { Media (idMal: $id, type: ANIME) { id } }`;
        const res = await fetch('https://graphql.anilist.co', {
            method: 'POST',
            headers: { 'Content-Type': 'application/json', 'Accept': 'application/json' },
            body: JSON.stringify({ query, variables: { id: parseInt(malId) } })
        });
        if (res.ok) {
            const data = await res.json() as any;
            const id = data.data?.Media?.id?.toString();
            if (id) {
                anilistCache.set(malId, id);
                return id;
            }
        }
    } catch (e) { console.error(`[Anilist] ID fetch error:`, e); }
    return null;
}

export async function searchAnimelok(query: string) {
    try {
        // Try suggestions first
        const sUrl = `https://animelok.xyz/api/anime/search-suggestions?q=${encodeURIComponent(query)}`;
        const sRes = await fetch(sUrl, { headers: ANIMELOK_HEADERS });
        if (sRes.ok) {
            const data = await sRes.json() as any;
            if (data.results?.length > 0) return data.results;
            if (data.length > 0) return data;
        }

        // Main search API
        const url = `https://animelok.xyz/api/anime/search?q=${encodeURIComponent(query)}`;
        const res = await fetch(url, { headers: ANIMELOK_HEADERS });
        if (res.ok) {
            const data = await res.json() as any;
            return data.results || data.anime || data || [];
        }
    } catch (e) { console.error(`[Animelok Search Error]:`, e); }
    return [];
}

export async function getAnimelokMetadata(slug: string) {
    let allEpisodes: any[] = [];
    let page = 0;
    const pageSize = 250;
    
    try {
        while (page < 10) { // Safety limit of 2500 episodes
            const url = `https://animelok.xyz/api/anime/${slug}/episodes-range?page=${page}&pageSize=${pageSize}`;
            const res = await fetch(url, { headers: ANIMELOK_HEADERS });
            if (!res.ok) break;
            
            const data = await res.json() as any;
            const eps = (data.episodes || []).map((ep: any) => ({
                number: ep.number,
                title: ep.name,
                image: ep.img,
                description: ep.description,
                isFiller: ep.isFiller
            }));
            
            if (eps.length === 0) break;
            allEpisodes.push(...eps);
            
            // If we got fewer than pageSize, it's likely the last page
            if (eps.length < pageSize) break;
            page++;
        }
        return allEpisodes;
    } catch (e) { console.error('[Animelok Metadata Error]', e); }
    return allEpisodes;
}

export async function getAnimelokSources(slug: string, ep: number) {
    try {
        const url = `https://animelok.xyz/api/anime/${slug}/episodes/${ep}`;
        const res = await fetch(url, { headers: ANIMELOK_HEADERS });
        if (res.ok) {
            const data = await res.json() as any;
                const servers = data.episode?.servers || [];
            
            const sources = servers.map((s: any) => {
                const tip = s.tip?.toLowerCase() || "";
                const name = s.name?.toLowerCase() || "";
                const languages = s.languages || [];
                
                let category = "sub";
                if (languages.includes("HINDI") || name.includes("hindi")) category = "hindi";
                else if (languages.includes("ENGLISH") && tip.includes("dub")) category = "dub";
                else if (languages.length > 1) category = "multi";
                
                let finalUrl = s.url || "";
                if (finalUrl.startsWith("[") && finalUrl.endsWith("]")) {
                    try {
                        const parsed = JSON.parse(finalUrl);
                        if (parsed && parsed.length > 0 && parsed[0].url) {
                            finalUrl = parsed[0].url;
                        }
                    } catch (e) {
                        // ignore parse error if not valid json
                    }
                }
                
                return {
                    name: s.name + (s.tip ? ` (${s.tip})` : ""),
                    url: finalUrl,
                    category: category,
                    language: languages[0] || (category === "dub" ? "English" : "Japanese"),
                    isM3U8: finalUrl.includes(".m3u8") || finalUrl.includes("master.m3u8") || finalUrl.includes("uwu.m3u8")
                };
            });
            
            const subtitlesList = data.episode?.subtitles || [];
            const parsedSubtitles = subtitlesList.map((sub: any) => ({
                language: sub.name || "Unknown",
                url: sub.url
            }));

            return { sources, subtitles: parsedSubtitles };
        }
    } catch (e) { console.error('[Animelok Sources Error]', e); }
    return { sources: [], subtitles: [] };
}

File: backend-go/internal/providers/animelok.go
package providers

import (
	"encoding/json"
	"fmt"
	"strings"

	"github.com/Ansh7473/anime-pro/backend-go/internal/utils"
)

const ANIMELOK_BASE = "https://animelok.xyz"

// SearchAnimelok performs a search on Animelok
func SearchAnimelok(query string) []map[string]interface{} {
	url := fmt.Sprintf("%s/api/anime/search?q=%s", ANIMELOK_BASE, utils.ToUrlQuery(query))
	resp, err := utils.HttpClient.R().
		SetHeader("Referer", ANIMELOK_BASE+"/").
		Get(url)
	
	if err != nil || !resp.IsSuccess() {
		return nil
	}

	var data map[string]interface{}
	if err := json.Unmarshal(resp.Body(), &data); err != nil {
		return nil
	}

	// Animelok API sometimes returns "results", sometimes "anime"
	results, _ := data["results"].([]interface{})
	if results == nil {
		results, _ = data["anime"].([]interface{})
	}

	transformed := make([]map[string]interface{}, 0)
	for _, item := range results {
		if m, ok := item.(map[string]interface{}); ok {
			transformed = append(transformed, m)
		}
	}
	return transformed
}

// GetAnimelokMetadataByMalId resolves slugs and fetches metadata with fallback
func GetAnimelokMetadataByMalId(malId int, titles []string) []map[string]interface{} {
	if len(titles) == 0 {
		return nil
	}

	// 1. Resolve IDs
	aniId, _ := utils.GetAnilistId(malId)
	
	baseSlug := strings.ToLower(titles[0])
	baseSlug = strings.ReplaceAll(baseSlug, " ", "-")
	// Clean slug
	reg := strings.NewReplacer("(", "", ")", "", ":", "", "!", "", "?", "", ".", "")
	baseSlug = reg.Replace(baseSlug)

	candidates := []string{
		fmt.Sprintf("%s-%d", baseSlug, malId),
	}
	if aniId > 0 && aniId != malId {
		candidates = append(candidates, fmt.Sprintf("%s-%d", baseSlug, aniId))
	}
	candidates = append(candidates, baseSlug)

	// Try candidates
	for _, slug := range candidates {
		eps := GetAnimelokMetadata(slug)
		if len(eps) > 0 {
			return eps
		}
	}

	// 2. Search Fallback
	searchResults := SearchAnimelok(titles[0])
	for _, res := range searchResults {
		slug := utils.ToString(res["slug"])
		if slug == "" {
			slug = utils.ToString(res["id"])
		}
		if slug != "" {
			eps := GetAnimelokMetadata(slug)
			if len(eps) > 0 {
				return eps
			}
		}
	}

	return nil
}


// GetAnimelokMetadata fetches all episodes for a given slug
func GetAnimelokMetadata(slug string) []map[string]interface{} {
	allEpisodes := make([]map[string]interface{}, 0)
	pageSize := 250

	for page := 0; page < 10; page++ {
		url := fmt.Sprintf("%s/api/anime/%s/episodes-range?page=%d&pageSize=%d", ANIMELOK_BASE, slug, page, pageSize)
		resp, err := utils.HttpClient.R().
			SetHeader("Referer", ANIMELOK_BASE+"/").
			Get(url)
		
		if err != nil || !resp.IsSuccess() {
			break
		}

		var data map[string]interface{}
		if err := json.Unmarshal(resp.Body(), &data); err != nil {
			break
		}

		eps, ok := data["episodes"].([]interface{})
		if !ok || len(eps) == 0 {
			break
		}

		for _, item := range eps {
			if ep, ok := item.(map[string]interface{}); ok {
				imageUrl := utils.ToString(ep["img"])
				if imageUrl != "" && !strings.HasPrefix(imageUrl, "http") {
					imageUrl = ANIMELOK_BASE + imageUrl
				}
				allEpisodes = append(allEpisodes, map[string]interface{}{
					"number": ep["number"],
					"title": ep["name"],
					"image": imageUrl,
					"description": ep["description"],
					"isFiller": ep["isFiller"],
				})
			}
		}

		if len(eps) < pageSize {
			break
		}
	}
	return allEpisodes
}

// GetAnimelokSources fetches streaming sources and subtitles for a given slug and episode
func GetAnimelokSources(slug string, ep int) map[string]interface{} {
	url := fmt.Sprintf("%s/api/anime/%s/episodes/%d", ANIMELOK_BASE, slug, ep)
	resp, err := utils.HttpClient.R().
		SetHeader("Referer", ANIMELOK_BASE+"/").
		Get(url)
	
	if err != nil || !resp.IsSuccess() {
		return map[string]interface{}{"sources": []interface{}{}, "subtitles": []interface{}{}}
	}

	var data map[string]interface{}
	if err := json.Unmarshal(resp.Body(), &data); err != nil {
		return map[string]interface{}{"sources": []interface{}{}, "subtitles": []interface{}{}}
	}

	episodeData, ok := data["episode"].(map[string]interface{})
	if !ok {
		return map[string]interface{}{"sources": []interface{}{}, "subtitles": []interface{}{}}
	}

	servers, _ := episodeData["servers"].([]interface{})
	sources := make([]map[string]interface{}, 0)
	for _, item := range servers {
		if s, ok := item.(map[string]interface{}); ok {
			name := strings.ToLower(utils.ToString(s["name"]))
			tip := strings.ToLower(utils.ToString(s["tip"]))
			
			// Detect category and language
			category := "sub"
			if strings.Contains(name, "hindi") || strings.Contains(tip, "hindi") {
				category = "hindi"
			} else if strings.Contains(tip, "dub") {
				category = "dub"
			}

			finalUrl := utils.ToString(s["url"])
			// Handle JSON-encoded URLs occasionally returned by some providers
			if strings.HasPrefix(finalUrl, "[") && strings.HasSuffix(finalUrl, "]") {
				var parsed []map[string]interface{}
				if err := json.Unmarshal([]byte(finalUrl), &parsed); err == nil && len(parsed) > 0 {
					finalUrl = utils.ToString(parsed[0]["url"])
				}
			}

			isM3U8 := strings.Contains(strings.ToLower(finalUrl), ".m3u8") || strings.Contains(strings.ToLower(finalUrl), "/m3u8")
			isEmbed := !isM3U8

			sources = append(sources, map[string]interface{}{
				"name":     utils.ToString(s["name"]) + " " + utils.ToString(s["tip"]),
				"url":      finalUrl,
				"category": category,
				"isM3U8":   isM3U8,
				"isEmbed":  isEmbed,
				"type":     "iframe",
				"provider": "Animelok",
			})
		}
	}

	subs, _ := episodeData["subtitles"].([]interface{})
	subtitles := make([]map[string]interface{}, 0)
	for _, item := range subs {
		if s, ok := item.(map[string]interface{}); ok {
			subtitles = append(subtitles, map[string]interface{}{
				"language": s["name"],
				"url": s["url"],
			})
		}
	}

	return map[string]interface{}{
		"sources": sources,
		"subtitles": subtitles,
	}
}
